Starter project

dlin96 opened the pull request:
Inconsistent key fix for writing tabular records.

krzentner commented on tests/dowel/test_csv_output.py:
```diff
@@ -35,6 +35,7 @@ def test_record(self):
         ]  # yapf: disable
         self.assert_csv_matches(correct)
 
+    @pytest.mark.skip(reason="conflicting functionality (overwritten for this feature)")
```
You can just delete the old test, it'll still be in the git history if we need it :)

krzentner commented on tests/dowel/test_csv_output.py:
```diff
@@ -85,11 +86,41 @@ def test_disable_warnings(self):
 
             for correct_row in correct:
                 row = next(reader)
+                print(row)
```
This is probably only here on accident?

krzentner commented on src/dowel/csv_output.py:
```diff
@@ -42,13 +43,36 @@ def record(self, data, prefix=''):
+
+                    for row in prev_reader:
+                        self._writer.writerow(row)
+                """
```
You can just delete the old code, it's already in the git history :)

We also generally avoid using multi-line strings as comments, it's confusing.

krzentner commented on src/dowel/csv_output.py:
```diff
@@ -42,13 +43,36 @@ def record(self, data, prefix=''):
+                    self._fieldnames.add(key)
+                
+                with open(self.output_file, "r") as orig:
+                    prev_reader = csv.DictReader(orig)
```
This reader reads lines on-demand, so it's not clear that opening the reader here before overwriting the file works in general.

krzentner commented on src/dowel/csv_output.py:
```diff
@@ -42,13 +43,36 @@ def record(self, data, prefix=''):
+                    prev_reader = csv.DictReader(orig)
+
+                    # need to create a new DictWriter object with new fieldnames
+                    self._writer = csv.DictWriter(
```
It's probably best to refactor out the writer initialization logic, so that if the parameters get changed somewhere they're consistently changed.

krzentner commented on src/dowel/csv_output.py:
```diff
@@ -42,13 +43,36 @@ def record(self, data, prefix=''):
+                    self._writer = csv.DictWriter(
+                        self._log_file,
+                        fieldnames=sorted(list(self._fieldnames)),
+                        extrasaction='ignore')
```
Why use `extrasaction='ignore'` here? It seems like `extrasaction='raise'` would be a safer choice.

dlin96 commented:
@krzentner 

Thank you so much for your comments! I believe I have fixed some of the formatting issues like removing unnecessary tests and print statements. Regarding the logic of the fix, I believe you are correct and that overwriting a file that's being read via generator it isn't guaranteed to be safe.

My best attempt at this would be to create a buffer file to store the data, then copy the data back over once the new writer is created. I believe I have refactored the CSV DictWriter as you mentioned, but please let me know if there's anything else I can improve!
